src/govuk/i18n.js

```diff
@@ -0,0 +1,418 @@
+import './vendor/polyfills/Array/prototype/indexOf'
+import './vendor/polyfills/Object/keys'
+
+/**
+ * i18n support initialisation function
+ *
+ * @constructor
+ * @param    {object}  options               - Options object.
+ * @param    {string}  options.locale        - The locale to use, currently this is only used for pluralisation rules. If undefined, tries to use html[lang].
+ * @param    {object}  options.translations  - Key-value pairs of the translation strings to use.
+ */
+function I18n (options) {
+  options = options || {}
+
+  // Get user-defined locale setting,
+  // Otherwise, try and detect it from the document,
+  // Otherwise, set to English
+  this.locale = options.locale || document.documentElement.lang || 'en'
+
+  // Define the separators used for placeholders.
+  // Currently uses the same syntax as GOV.UK (e.g. %{keyName}).
+  // e.g. https://github.com/alphagov/govuk_publishing_components/blob/22944a9efaea3ae495a518717415694c71cfdb35/config/locales/en.yml
+  this.separators = ['%{', '}']
+
+  // Flatten the list of translations
+  this.translations = options.translations ? this.flatten(options.translations) : {}
+
+  if (!I18n._instance) {
+    I18n._instance = this
+  }
+}
+
+/**
+ * Static function to get the "default" (currently, always the first) instance
+ * of the I18n function. If no default instance exists, a new one is created
+ * with the default options.
+ *
+ * @static
+ * @returns  { function }  - A configured instance of the I18n function.
+ */
+I18n.getInstance = function () {
+  if (I18n._instance) {
+    return I18n._instance
+  }
+  return new I18n()
+}
+
+/**
+ * The most used function - looks up translation strings and sorts what
+ * operations to perform on them.
+ *
+ * ALL properties passed to `options` will be exposed as options for string
+ * interpolation, but only `count` and `fallback` initiate special functionality.
+ *
+ * TODO: Do we want to put string interpolation options in their own object (e.g. `options.data`) to avoid potential conflict in future?
```

Comment:
Useful from the perspective of users not introducing breaking changes in the future, though p'raps a bit more mental load for users to take on right now. I wouldn't be against this.